wikimedia/wikipedia-ios
# Talk Page Archives - Part 1

**Phabricator:** 
https://phabricator.wikimedia.org/T321853

### Notes
This is some foundational work for talk page archives. I've been wanting to use this simple list feature as an opportunity to tinker with alternative navigation and header views for SwiftUI. Our custom navigation setup (currently in ViewController.swift) is heavily tied to UIKit scroll views, so in order to lean on SwiftUI ScrollViews, I wrote a different setup. Here's loosley how it works, but I'm happy to explain further in PR review if needed.

- The SwiftUI view is added to a `UIHostingController`, which in turn is added as a child view controller to `TalkPageArchivesViewController`
- `TalkPageArchivesViewController` also adds a `UIStackView` of the at the top of its view. This stack view holds the navigation bar and header view.
- There's a shared observable object called `ShiftingTopViewsData` that holds the scroll view offset and total height of the stack view. SwiftUI updates the scroll view offset in `ShiftingTopViewsData`, and the stack view listens to that value and adjusts its subviews accordingly. In turn the stack view updates its total height in `ShiftingTopViewsData`, and the SwiftUI scroll view listens to that and adds top padding accordingly.

I have added some mock UI to play with - I will delete it in the next PR.

I went with a super literal "ShiftingTopView" prefix everywhere, but happy to change this if it sounds weird. This was partly to distinguish it from the old system, and partly to avoid anything that sounded like a `UINavigationController` subclass to reduce confusion.

If you're interested in where this is headed, some prototype work is in the `talk-page-archives-prototype` branch. There I have it implemented on `TalkPageViewController`, and confirmed the sticky header issue is gone :).

### Test Steps
1. Set theme to "default". Go to a talk page.
2. In the overflow menu tap "Archives"
3. Scroll around, rotate, confirm layout bugs occur. Background and change the device dark mode toggle. Return and confirm all views update theming.
